keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+indicates whether a driver deployment will create `CSIStorageCapacity`
+objects with capacity information and wants the Kubernetes scheduler
+to rely on that information when making scheduling decisions that
+involve volumes that need to be created by the driver.
+
+If not set or false, the scheduler makes such decisions without considering
+whether the driver really can create the volumes (the current situation).
+
+### Updating capacity information with external-provisioner
+
+Most (if not all) CSI drivers already get deployed on Kubernetes
+together with the external-provisioner which then handles volume
+provisioning via PVC.
+
+Because the external-provisioner is part of the deployment of the CSI
+driver, that deployment can configure the behavior of the
+external-provisioner via command line parameters. There is no need to
+introduce heuristics or other, more complex ways of changing the
+behavior (like extending the `CSIDriver` API).
+
+#### Available capacity vs. maximum volume size
+
+The CSI spec up to and including the current version 1.2 just
+specifies that ["the available
+capacity"](https://github.com/container-storage-interface/spec/blob/314ac542302938640c59b6fb501c635f27015326/lib/go/csi/csi.pb.go#L2548-L2554)
+is to be returned by the driver. It is left open whether that means
+that a volume of that size can be created. This KEP uses the reported
+capacity to rule out pools which clearly have insufficient storage
+because the reported capacity is smaller than the size of a
+volume. This will work better when CSI drivers implement `GetCapacity`
+such that they consider constraints like fragmentation and report the
+size that the largest volume can have at the moment.
+
+#### Without central controller
+
+This mode of operation is expected to be used by CSI drivers that need
+to track capacity per node and only support persistent volumes with
+delayed binding that get provisioned by an external-provisioner
+instance that runs on the node where the volume gets provisioned (see
+the proposal for csi-driver-host-path in
+https://github.com/kubernetes-csi/external-provisioner/pull/367).
+
+The CSI driver has to implement the CSI controller service and its
+`GetCapacity` call. Its deployment has to add the external-provisioner
+to the daemon set and enable the per-node capacity tracking with
+`--enable-capacity=local`.
+
+The resulting `CSIStorageCapacity` objects then use a node selector for
+one node.
+
+#### With central controller
+
+For central provisioning, external-provisioner gets deployed together
+with a CSI controller service and capacity reporting gets enabled with
+`--enable-capacity=central`. In this mode, CSI drivers must report
+topology information in `NodeGetInfoResponse.accessible_topology` that
+matches the storage pool(s) that it has access to, with granularity
+that matches the most restrictive pool.
+
+For example, if the driver runs in a node with zone/region/rack
```

Comment:
Suppose the storage is local to a region. There's storage with label region=west and storage with region=east. Now in each region, we have data centers. As long as NodeGetInfo returns region=east resp. region=west for each node depending on where it is located, the proposed algorithm works.

Now suppose that in addition to storage in each region, each data center also has storage. Some storage classes select the region storage, some others the faster storage in the datacenter.

The storage in a New York data center then has "region=east,center=NewYork" as topology, and NodeGetInfo for a node in that center has to return that.

The proposed algorithm will then still work: it will use region+center for the topologies and create `CSIStorageCapacity` objects with that topology also for region storage.

What isn't okay is when NodeGetInfo only reports region, because that is not detailed enough for the storage in the centers.

Reply:
Thanks for clarifying. I think the original example is confusing because it seems like zone encompasses a region, which is the reverse of what most cloud providers do. Can we change the example to use a more clear hierarchy, like region and rack?